File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/ClassLoaderResourceLocation.java
package com.greentree.commons.data.resource.location;

import java.net.URL;
import java.util.Objects;

import com.greentree.commons.data.resource.URLResource;

public class ClassLoaderResourceLocation implements ResourceLocation {
	private static final long serialVersionUID = 1L;
	
	private final Class<?> cls;

	public ClassLoaderResourceLocation() {
		this(ClassLoaderResourceLocation.class);
	}
	public ClassLoaderResourceLocation(Class<?> cls) {
		this.cls = cls;
	}

	@Override
	public boolean equals(Object obj) {
		if(this == obj) return true;
		if(obj == null || getClass() != obj.getClass()) return false;
		ClassLoaderResourceLocation other = (ClassLoaderResourceLocation) obj;
		return Objects.equals(cls, other.cls);
	}

	public ClassLoader getLoader() {
		return cls.getClassLoader();
	}

	@Override
	public URLResource getResource(String name) {
		final var url = getURL(name);
		if(url == null) return null;
		return new URLResource(url);
	}

	public URL getURL(String name) {
		if(name.isBlank()) return null;
		return getLoader().getResource(name);
	}

	@Override
	public int hashCode() {
		return Objects.hash(cls);
	}

	@Override
	public String toString() {
		return "ClassLoaderResourceLocation [" + cls + "]";
	}



}

File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/ClassLoaderResourceLocation.kt
package com.greentree.commons.data.resource.location

import com.greentree.commons.data.resource.FileResource
import com.greentree.commons.data.resource.Resource
import com.greentree.commons.data.resource.URLFileResource
import java.net.URL

data class ClassLoaderResourceLocation(val loader: ClassLoader) : ResourceLocation {

	constructor(cls: Class<*> = ClassLoaderResourceLocation::class.java) : this(cls.classLoader)

	override fun getResourceOrNull(name: String): Resource? {
		TODO("Not yet implemented")
	}

	override fun getFileResourceOrNull(name: String): FileResource? {
		val url = getURL(name) ?: return null
		return URLFileResource(url)
	}

	private fun getURL(name: String): URL? {
		return loader.getResource(name)
	}

	override fun toString(): String {
		return "ClassLoaderResourceLocation [$loader]"
	}
}